microgateway_decorator/bin/compile

```diff
@@ -7,12 +7,95 @@ BUILD_DIR=$1
+# or the url is missing and the file name is present
+# or the url is present and the file name is missing
+function check_nodejs_filename_and_url {
+  echo "Checking APIGEE_MICROGATEWAY_NODEJS_FILENAME and APIGEE_MICROGATEWAY_NODEJS_URL environment variables."
+  # check if the value is set to false or if it is zero length then assume false.
+  if [ "$APIGEE_MICROGATEWAY_NODEJS_LOCAL_INSTALL" == false ] || [ -z "$APIGEE_MICROGATEWAY_NODEJS_LOCAL_INSTALL" ]; then
+    echo "APIGEE_MICROGATEWAY_NODEJS_LOCAL_INSTALL is either set to false or assumed false."
+    if [ "$APIGEE_MICROGATEWAY_NODEJS_URL" != "" ] && [ "$APIGEE_MICROGATEWAY_NODEJS_FILENAME" == "" ]; then
+      echo "If you set APIGEE_MICROGATEWAY_NODEJS_URL=$APIGEE_MICROGATEWAY_NODEJS_URL then you must also set APIGEE_MICROGATEWAY_NODEJS_FILENAME as well." >&2
+      exit 1
+    elif [ "$APIGEE_MICROGATEWAY_NODEJS_URL" == "" ] && [ "$APIGEE_MICROGATEWAY_NODEJS_FILENAME" != "" ]; then
+      echo "If you set APIGEE_MICROGATEWAY_NODEJS_FILENAME=$APIGEE_MICROGATEWAY_NODEJS_FILENAME then you must also set APIGEE_MICROGATEWAY_NODEJS_URL as well." >&2
+      exit 1
+    else
+      if [ "$APIGEE_MICROGATEWAY_NODEJS_FILENAME" != "" ]; then
+        if ! grep -q "http://" <<< "$APIGEE_MICROGATEWAY_NODEJS_URL" && ! grep -q "https://" <<< "$APIGEE_MICROGATEWAY_NODEJS_URL" ; then
+          echo "$APIGEE_MICROGATEWAY_NODEJS_URL does not include a valid protocol (https or http)" >&2
+          exit 1
+        fi
+      fi
+    fi
+  elif [ $APIGEE_MICROGATEWAY_NODEJS_LOCAL_INSTALL == true ]; then
+    if [ ! -f "$ROOT_DIR/lib/$APIGEE_MICROGATEWAY_NODEJS_FILENAME" ]; then
+      echo "APIGEE_MICROGATEWAY_NODEJS_LOCAL_INSTALL is set to true but APIGEE_MICROGATEWAY_NODEJS_FILENAME=$APIGEE_MICROGATEWAY_NODEJS_FILENAME and that file does not exist in the lib directory. Exiting..." >&2
+      exit 1
+    fi
+  fi
+  echo "APIGEE_MICROGATEWAY_NODEJS_FILENAME and APIGEE_MICROGATEWAY_NODEJS_URL environment variables are set correctly."
+}
+
 mkdir -p $BUILD_DIR/tmp/node
 mkdir -p $BUILD_DIR/tmp/edgemicro
 
-wget --quiet https://nodejs.org/dist/v6.11.3/node-v6.11.3-linux-x64.tar.gz
-tar --touch --strip-components 1 -xzf node-v6.11.3-linux-x64.tar.gz -C $BUILD_DIR/tmp/node
-rm node-v6.11.3-linux-x64.tar.gz
+check_nodejslocalinstall_env_var
+check_nodejs_filename_and_url
+
+if [ "$APIGEE_MICROGATEWAY_NODEJS_LOCAL_INSTALL" == true ]; then
+  echo "This is a local Node.js installation."
+  untar "$ROOT_DIR/lib/$APIGEE_MICROGATEWAY_NODEJS_FILENAME"
+elif [ ! -z "$APIGEE_MICROGATEWAY_NODEJS_URL" ]; then
+  echo "Fetch Node.js tar from $APIGEE_MICROGATEWAY_NODEJS_URL/$APIGEE_MICROGATEWAY_NODEJS_FILENAME"
+  wget_and_untar "$APIGEE_MICROGATEWAY_NODEJS_URL/$APIGEE_MICROGATEWAY_NODEJS_FILENAME" "$APIGEE_MICROGATEWAY_NODEJS_FILENAME"
+elif [ -z "$APIGEE_MICROGATEWAY_NODEJS_VERSION" ]; then
+  echo "Fetch Node.js from https://nodejs.org/dist/v8.11.3/node-v8.11.3-linux-x64.tar.gz"
```

Comment:
small nit. Lets make this a variable instead of hard coding

Reply:
I'll change this.